File: modules/slicot/src/fortran/MB03AD.f
      SUBROUTINE MB03AD( SHFT, K, N, AMAP, S, SINV, A, LDA1, LDA2, C1,
     $                   S1, C2, S2 )
C
C     PURPOSE
C
C     To compute two Givens rotations (C1,S1) and (C2,S2) such that the
C     orthogonal matrix
C
C               [ Q  0 ]        [  C1  S1  0 ]   [ 1  0   0  ]
C           Z = [      ],  Q := [ -S1  C1  0 ] * [ 0  C2  S2 ],
C               [ 0  I ]        [  0   0   1 ]   [ 0 -S2  C2 ]
C
C     makes the first column of the real Wilkinson double shift
C     polynomial of the product of matrices in periodic upper Hessenberg
C     form, stored in the array A, parallel to the first unit vector.
C     Only the rotation defined by C1 and S1 is used for the real
C     Wilkinson single shift polynomial (see SLICOT Library routine
C     MB03BE).
C
C     ARGUMENTS
C
C     Mode Parameters
C
C     SHFT    CHARACTER*1
C             Specifies the number of shifts employed by the shift
C             polynomial, as follows:
C             = 'D':  two shifts (assumes N > 2);
C             = 'S':  one real shift.
C
C     Input/Output Parameters
C
C     K       (input)  INTEGER
C             The number of factors.  K >= 1.
C
C     N       (input)  INTEGER
C             The order of the factors in the array A.
C             N >= 2, for a single shift polynomial;
C             N >= 3, for a double shift polynomial.
C
C     AMAP    (input)  INTEGER array, dimension (K)
C             The map for accessing the factors, i.e., if AMAP(I) = J,
C             then the factor A_I is stored at the J-th position in A.
C             AMAP(1) is the pointer to the Hessenberg matrix.
C
C     S       (input)  INTEGER array, dimension (K)
C             The signature array. Each entry of S must be 1 or -1.
C
C     SINV    (input)  INTEGER
C             Signature multiplier. Entries of S are virtually
C             multiplied by SINV.
C
C     A       (input)  DOUBLE PRECISION array, dimension (LDA1,LDA2,K)
C             The leading N-by-N-by-K part of this array must contain
C             the product (implicitly represented by its K factors)
C             in periodic upper Hessenberg form.
C
C     LDA1    INTEGER
C             The first leading dimension of the array A.  LDA1 >= N.
C
C     LDA2    INTEGER
C             The second leading dimension of the array A.  LDA2 >= N.
C
C     C1      (output)  DOUBLE PRECISION
C     S1      (output)  DOUBLE PRECISION
C             On exit, C1 and S1 contain the parameters for the first
C             Givens rotation.
C
C     C2      (output)  DOUBLE PRECISION
C     S2      (output)  DOUBLE PRECISION
C             On exit, if SHFT = 'D' and N > 2, C2 and S2 contain the
C             parameters for the second Givens rotation. Otherwise,
C             C2 = 1, S2 = 0.
C
C     METHOD
C
C     Two Givens rotations are properly computed and applied.
C
C     CONTRIBUTOR
C
C     D. Kressner, Technical Univ. Berlin, Germany, June 2001.
C
C     REVISIONS
C
C     V. Sima, Research Institute for Informatics, Bucharest, Romania,
C     July 2009, SLICOT Library version of the routine PLASHF.
C     V. Sima, Apr. 2018, Oct. 2019, Dec. 2019.
C
C     KEYWORDS
C
C     Eigenvalues, QZ algorithm, periodic QZ algorithm, orthogonal
C     transformation.
C
C     ******************************************************************
C
C     .. Parameters ..
      DOUBLE PRECISION  ZERO, ONE, TWO
      PARAMETER         ( ZERO = 0.0D0, ONE = 1.0D0, TWO = 2.0D0 )
C     .. Scalar Arguments ..
      CHARACTER         SHFT
      INTEGER           K, LDA1, LDA2, N, SINV
      DOUBLE PRECISION  C1, C2, S1, S2
C     .. Array Arguments ..
      INTEGER           AMAP(*), S(*)
      DOUBLE PRECISION  A(LDA1,LDA2,*)
C     .. Local Scalars ..
      LOGICAL           SGLE
      INTEGER           AI, I
      DOUBLE PRECISION  ALPHA, BETA, C3, DELTA, GAMMA, S3, TEMP
C     .. External Functions ..
      LOGICAL           LSAME
      EXTERNAL          LSAME
C     .. External Subroutines ..
      EXTERNAL          DLARTG
C     .. Intrinsic Functions ..
      INTRINSIC         SQRT
C
C     .. Executable Statements ..
C
      SGLE = LSAME( SHFT, 'S' )
      C1 = ONE
      S1 = ZERO
      C2 = 1/SQRT( TWO )
      S2 = C2
C
      DO 10  I = K, 2, -1
         AI = AMAP(I)
         IF ( S(AI).EQ.SINV ) THEN
            ALPHA = C2 * A(1,1,AI)
            GAMMA = S2 * A(N,N,AI)
            BETA  = S2 * A(N-1,N,AI)
            BETA  = C1 * BETA + S1 * A(N-1,N-1,AI)
            CALL DLARTG( ALPHA, GAMMA, C2, S2, TEMP )
            TEMP = C1 * TEMP
            CALL DLARTG( TEMP, BETA, C1, S1, ALPHA )
         ELSE
            TEMP  = A(1,1,AI)
            BETA  = S2 * TEMP
            TEMP  = C2 * TEMP
            ALPHA = S1 * TEMP
            GAMMA = A(N,N,AI)
            DELTA = C2 * GAMMA
            GAMMA = S2 * GAMMA
            CALL DLARTG( DELTA, BETA, C2, S2, C3 )
            DELTA = C1 * A(N-1,N,AI) - S1 * GAMMA
            ALPHA = C2 * ALPHA - S2 * DELTA
            GAMMA = C1 * A(N-1,N-1,AI)
            CALL DLARTG( GAMMA, ALPHA, C1, S1, TEMP )
         END IF
   10 CONTINUE
C
      AI = AMAP(1)
      ALPHA = A(1,1,AI) * C2 - A(N,N,AI) * S2
      BETA  = C1 * ( C2 * A(2,1,AI) )
      GAMMA = C1 * ( S2 * A(N-1,N,AI) ) + S1 * A(N-1,N-1,AI)
      ALPHA = ALPHA * C1 - A(N,N-1,AI) * S1
      CALL DLARTG( ALPHA, BETA, C1, S1, TEMP )
C
C     This is sufficient for a single real shift.
C
      IF ( SGLE ) THEN
         C2 = ONE
         S2 = ZERO
C
      ELSE
C
         CALL DLARTG( TEMP, GAMMA, C2, S2, ALPHA )
C
C        Rotation 1 is preserved.
C
         ALPHA = C2
         GAMMA = ( A(N-1,N-1,AI) * C1 ) * C2 + A(N,N-1,AI) * S2
         DELTA = ( A(N-1,N-1,AI) * S1 ) * C2
         CALL DLARTG( GAMMA, DELTA, C3, S3, TEMP )
         CALL DLARTG( ALPHA, TEMP, C2, S2, ALPHA )
C
C        Rotation 3 is preserved throughout the following complete loop.
C
         DO 20  I = K, 2, -1
            AI = AMAP(I)
            IF ( S(AI).EQ.SINV ) THEN
               ALPHA = ( A(1,1,AI) * C1 + A(1,2,AI) * S1 ) * C2
               BETA  = ( A(2,2,AI) * S1 ) * C2
               GAMMA = A(N-1,N-1,AI) * S2
               CALL DLARTG( ALPHA, BETA, C1, S1, TEMP )
               CALL DLARTG( TEMP, GAMMA, C2, S2, ALPHA )
            ELSE
               ALPHA =  C1 * A(1,1,AI)
               GAMMA =  S1 * A(1,1,AI)
               BETA  =  C1 * A(1,2,AI) + S1 * A(2,2,AI)
               DELTA = -S1 * A(1,2,AI) + C1 * A(2,2,AI)
               CALL DLARTG( DELTA, GAMMA, C1, S1, TEMP )
               ALPHA = -ALPHA * S2
               BETA  = -BETA  * S2
               ALPHA = C1 * ALPHA + S1 * BETA
               BETA  = C2 * A(N-1,N-1,AI)
               CALL DLARTG( BETA, ALPHA, C2, S2, TEMP )
               S2 = -S2
            END IF
   20    CONTINUE
C
C        Last step: Let the rotations collap into A.
C
         AI = AMAP(1)
         ALPHA = C1 * A(1,1,AI) + S1 * A(1,2,AI)
         BETA  = C1 * A(2,1,AI) + S1 * A(2,2,AI)
         GAMMA = S1 * A(3,2,AI)
         ALPHA = C2 * ALPHA - S2 * C3
         BETA  = C2 * BETA  - S2 * S3
         GAMMA = C2 * GAMMA
         CALL DLARTG( BETA, GAMMA, C2, S2, TEMP )
         CALL DLARTG( ALPHA, TEMP, C1, S1, BETA )
      END IF
      RETURN
C *** Last line of MB03AD ***
      END

File: modules/slicot/src/c/MB03AD.c
/* Translated by Nelson f2c (version 20170901).
   You must link the resulting object file with the libraries:
    -lnlsf2c -lm   (in that order)
*/

#include "nelson_f2c.h"

EXPORTSYMBOL /* Subroutine */ int mb03ad_(
    shft, k, n, amap, s, sinv, a, lda1, lda2, c1, s1, c2, s2, shft_len) char* shft;
integer *k, *n, *amap, *s, *sinv;
doublereal* a;
integer *lda1, *lda2;
doublereal *c1, *s1, *c2, *s2;
ftnlen shft_len;
{
    /* System generated locals */
    integer a_dim1, a_dim2, a_offset;
    /* Builtin functions */
    double sqrt();
    /* Local variables */
    static doublereal beta;
    static logical sgle;
    static doublereal temp;
    static integer i__;
    static doublereal gamma, alpha, delta;
    extern logical lsame_();
    static doublereal c3, s3;
    static integer ai;
    extern /* Subroutine */ int dlartg_();
    /*     SLICOT RELEASE 5.0. */
    /*     Copyright (c) 2002-2010 NICONET e.V. */
    /*     This program is free software: you can redistribute it and/or */
    /*     modify it under the terms of the GNU General Public License as */
    /*     published by the Free Software Foundation, either version 2 of */
    /*     the License, or (at your option) any later version. */
    /*     This program is distributed in the hope that it will be useful, */
    /*     but WITHOUT ANY WARRANTY; without even the implied warranty of */
    /*     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the */
    /*     GNU General Public License for more details. */
    /*     You should have received a copy of the GNU General Public License */
    /*     along with this program.  If not, see */
    /*     <http://www.gnu.org/licenses/>. */
    /*     PURPOSE */
    /*     To compute two Givens rotations (C1,S1) and (C2,S2) */
    /*     such that the orthogonal matrix */
    /*                [  C1  S1  0 ]   [ 1  0   0  ] */
    /*           Q =  [ -S1  C1  0 ] * [ 0  C2  S2 ] */
    /*                [  0   0   1 ]   [ 0 -S2  C2 ] */
    /*     makes the first column of the real Wilkinson single/double shift */
    /*     polynomial of the general product of matrices, stored in the */
    /*     array A, parallel to the first unit vector. */
    /*     ARGUMENTS */
    /*     Mode Parameters */
    /*     SHFT    CHARACTER*1 */
    /*             Specifies the number of shifts employed by the shift */
    /*             polynomial, as follows: */
    /*             = 'D':  two real shifts; */
    /*             = 'S':  one real shift. */
    /*     Input/Output Parameters */
    /*     K       (input)  INTEGER */
    /*             The number of factors.  K >= 1. */
    /*     N       (input)  INTEGER */
    /*             The order of the factors in the array A.  N >= 3. */
    /*     AMAP    (input) INTEGER array, dimension (K) */
    /*             The map for accessing the factors, i.e., if AMAP(I) = J, */
    /*             then the factor A_I is stored at the J-th position in A. */
    /*     S       (input)  INTEGER array, dimension (K) */
    /*             The signature array. Each entry of S must be 1 or -1. */
    /*     SINV    (input) INTEGER */
    /*             Signature multiplier. Entries of S are virtually */
    /*             multiplied by SINV. */
    /*     A       (input)  DOUBLE PRECISION array, dimension (LDA1,LDA2,K) */
    /*             On entry, the leading N-by-N-by-K part of this array must */
    /*             contain a n-by-n product (implicitly represented by its K */
    /*             factors) in upper Hessenberg form. */
    /*     LDA1    INTEGER */
    /*             The first leading dimension of the array A.  LDA1 >= N. */
    /*     LDA2    INTEGER */
    /*             The second leading dimension of the array A.  LDA2 >= N. */
    /*     C1      (output)  DOUBLE PRECISION */
    /*     S1      (output)  DOUBLE PRECISION */
    /*             On exit, C1 and S1 contain the parameters for the first */
    /*             Givens rotation. */
    /*     C2      (output)  DOUBLE PRECISION */
    /*     S2      (output)  DOUBLE PRECISION */
    /*             On exit, if SHFT = 'D', C2 and S2 contain the parameters */
    /*             for the second Givens rotation. */
    /*     METHOD */
    /*     Two Givens rotations are properly computed and applied. */
    /*     CONTRIBUTOR */
    /*     D. Kressner, Technical Univ. Berlin, Germany, June 2001. */
    /*     REVISIONS */
    /*     V. Sima, Research Institute for Informatics, Bucharest, Romania, */
    /*     July 2009, SLICOT Library version of the routine PLASHF. */
    /*     KEYWORDS */
    /*     Eigenvalues, QZ algorithm, periodic QZ algorithm, orthogonal */
    /*     transformation. */
    /*     ****************************************************************** */
    /*     .. Parameters .. */
    /*     .. Scalar Arguments .. */
    /*     .. Array Arguments .. */
    /*     .. Local Scalars .. */
    /*     .. External Functions .. */
    /*     .. External Subroutines .. */
    /*     .. Intrinsic Functions .. */
    /*     .. Executable Statements .. */
    /* Parameter adjustments */
    --amap;
    --s;
    a_dim1 = *lda1;
    a_dim2 = *lda2;
    a_offset = a_dim1 * (a_dim2 + 1) + 1;
    a -= a_offset;
    /* Function Body */
    sgle = lsame_(shft, "S", 1L, 1L);
    *c1 = 1.;
    *s1 = 0.;
    *c2 = 1 / sqrt(2.);
    *s2 = *c2;
    for (i__ = *k; i__ >= 2; --i__) {
        ai = amap[i__];
        if (s[ai] == *sinv) {
            alpha = *c2 * a[(ai * a_dim2 + 1) * a_dim1 + 1];
            gamma = *s2 * a[*n + (*n + ai * a_dim2) * a_dim1];
            beta = *s2 * a[*n - 1 + (*n + ai * a_dim2) * a_dim1];
            beta = *c1 * beta + *s1 * a[*n - 1 + (*n - 1 + ai * a_dim2) * a_dim1];
            dlartg_(&alpha, &gamma, c2, s2, &temp);
            temp = *c1 * temp;
            dlartg_(&temp, &beta, c1, s1, &alpha);
        } else {
            temp = a[(ai * a_dim2 + 1) * a_dim1 + 1];
            beta = *s2 * temp;
            temp = *c2 * temp;
            alpha = *s1 * temp;
            gamma = a[*n + (*n + ai * a_dim2) * a_dim1];
            delta = *c2 * gamma;
            gamma = *s2 * gamma;
            dlartg_(&delta, &beta, c2, s2, &c3);
            delta = *c1 * a[*n - 1 + (*n + ai * a_dim2) * a_dim1] - *s1 * gamma;
            alpha = *c2 * alpha - *s2 * delta;
            gamma = *c1 * a[*n - 1 + (*n - 1 + ai * a_dim2) * a_dim1];
            dlartg_(&gamma, &alpha, c1, s1, &temp);
        }
        /* L10: */
    }
    ai = amap[1];
    alpha = a[(ai * a_dim2 + 1) * a_dim1 + 1] * *c2 - a[*n + (*n + ai * a_dim2) * a_dim1] * *s2;
    beta = *c1 * (*c2 * a[(ai * a_dim2 + 1) * a_dim1 + 2]);
    gamma = *c1 * (*s2 * a[*n - 1 + (*n + ai * a_dim2) * a_dim1])
        + *s1 * a[*n - 1 + (*n - 1 + ai * a_dim2) * a_dim1];
    alpha = alpha * *c1 - a[*n + (*n - 1 + ai * a_dim2) * a_dim1] * *s1;
    dlartg_(&alpha, &beta, c1, s1, &temp);
    /*     This is sufficient for single real shifts. */
    if (!sgle) {
        dlartg_(&temp, &gamma, c2, s2, &alpha);
        /*        Rotation 1 is preserved. */
        alpha = *c2;
        gamma = a[*n - 1 + (*n - 1 + ai * a_dim2) * a_dim1] * *c1 * *c2
            + a[*n + (*n - 1 + ai * a_dim2) * a_dim1] * *s2;
        delta = a[*n - 1 + (*n - 1 + ai * a_dim2) * a_dim1] * *s1 * *c2;
        dlartg_(&gamma, &delta, &c3, &s3, &temp);
        dlartg_(&alpha, &temp, c2, s2, &alpha);
        /*        Rotation 3 is preserved throughout the following complete loop. */
        for (i__ = *k; i__ >= 2; --i__) {
            ai = amap[i__];
            if (s[ai] == *sinv) {
                alpha = (a[(ai * a_dim2 + 1) * a_dim1 + 1] * *c1
                            + a[(ai * a_dim2 + 2) * a_dim1 + 1] * *s1)
                    * *c2;
                beta = a[(ai * a_dim2 + 2) * a_dim1 + 2] * *s1 * *c2;
                gamma = a[*n - 1 + (*n - 1 + ai * a_dim2) * a_dim1] * *s2;
                dlartg_(&alpha, &beta, c1, s1, &temp);
                dlartg_(&temp, &gamma, c2, s2, &alpha);
            } else {
                alpha = *c1 * a[(ai * a_dim2 + 1) * a_dim1 + 1];
                gamma = *s1 * a[(ai * a_dim2 + 1) * a_dim1 + 1];
                beta = *c1 * a[(ai * a_dim2 + 2) * a_dim1 + 1]
                    + *s1 * a[(ai * a_dim2 + 2) * a_dim1 + 2];
                delta = -(*s1) * a[(ai * a_dim2 + 2) * a_dim1 + 1]
                    + *c1 * a[(ai * a_dim2 + 2) * a_dim1 + 2];
                dlartg_(&delta, &gamma, c1, s1, &temp);
                alpha = -alpha * *s2;
                beta = -beta * *s2;
                alpha = *c1 * alpha + *s1 * beta;
                beta = *c2 * a[*n - 1 + (*n - 1 + ai * a_dim2) * a_dim1];
                dlartg_(&beta, &alpha, c2, s2, &temp);
                *s2 = -(*s2);
            }
            /* L20: */
        }
        /*        Last step: Let the rotations collap into A. */
        ai = amap[1];
        alpha = *c1 * a[(ai * a_dim2 + 1) * a_dim1 + 1] + *s1 * a[(ai * a_dim2 + 2) * a_dim1 + 1];
        beta = *c1 * a[(ai * a_dim2 + 1) * a_dim1 + 2] + *s1 * a[(ai * a_dim2 + 2) * a_dim1 + 2];
        gamma = *s1 * a[(ai * a_dim2 + 2) * a_dim1 + 3];
        alpha = *c2 * alpha - *s2 * c3;
        beta = *c2 * beta - *s2 * s3;
        gamma = *c2 * gamma;
        dlartg_(&beta, &gamma, c2, s2, &temp);
        dlartg_(&alpha, &temp, c1, s1, &beta);
    }
    return 0;
    /* *** Last line of MB03AD *** */
} /* mb03ad_ */
